[WiP] [DNM] Update power management support.

pizi-nordic opened the pull request:
This PR updates the nRF5x power management code after recent changes in the power management subsystem in Zephyr. It also updates the `power_mgr` example for nRF5x in order to show how to use existing power management policies. For details, please look into the commit descriptions.

@ramakrishnapallala:
While working on this PR, few ideas popped up in my mind:
- Each platform defines own `enum power_states` as well as global power management functions (like the `sys_set_power_state()` for example). Could we move these to global header as these definitions have to be the same on all platforms?
- I also think that  `sys_is_valid_power_state()` could be generic too, as we can determine which states are available basing on the Kconfig defines.

I have tried to write the header and the  `sys_is_valid_power_state()` function in the way which allow to reuse them in the generic code. If you think that it is a good idea, feel free to move this code out of the `soc` tree.

---- Note 1 ----
I have taken opportunity to implement things I mentioned above as well as some other improvements.
For the time I tagged this PR as DNM and WiP.

---- Note 2 ----
~~This PR is based on top of https://github.com/zephyrproject-rtos/zephyr/pull/12557~~ (https://github.com/zephyrproject-rtos/zephyr/pull/12557 is already merged)

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -9,30 +9,31 @@ Overview
-1. Low Power State: Low Power State is SOC specific and being in this state is
-   transparent to devices. SOC and devices do not lose context in this Mode.
-   SOC supports the following two Low Power states :
+1. Low Power State: Low Power State could be SoC, Board and/or Application
```
```suggestion
1. Low Power State: Low Power State could be SoC, board, or application
```

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -9,30 +9,31 @@ Overview
-   transparent to devices. SOC and devices do not lose context in this Mode.
-   SOC supports the following two Low Power states :
+1. Low Power State: Low Power State could be SoC, Board and/or Application
+   specific and being in this state is transparent to devices. SOC and devices
```
```suggestion
   specific. Being in this state is transparent to devices. SOC and devices
```

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -54,39 +55,38 @@ Running:
 
-  Press BUTTON1 to enter into Deep Dleep state...Press BUTTON2 to exit Deep Sleep state
-  --> Entering into Deep Sleep State -
+   Press BUTTON1 to enter into Deep Dleep state. Press BUTTON2 to exit Deep Sleep state.
```
```suggestion
   Press BUTTON1 to enter into Deep Sleep state. Press BUTTON2 to exit Deep Sleep state.
```

dbkinder commented on samples/boards/nrf52/power_mgr/src/main.c:
```diff
@@ -4,56 +4,76 @@
 	}
 
-	printk("\nPress BUTTON1 to enter into Deep Dleep state..."
+	printk("\nPress BUTTON1 to enter into Deep Dleep state. "
```
```suggestion
 	printk("\nPress BUTTON1 to enter into Deep Sleep state. "
```

nordic-krch commented on soc/arm/nordic_nrf/nrf52/power.c:
```diff
@@ -82,38 +32,63 @@ void sys_set_power_state(enum power_states state)
 {
 	switch (state) {
+#ifdef CONFIG_SYS_POWER_LOW_POWER_STATE
+# ifdef CONFIG_SYS_POWER_STATE_CPU_LPS_SUPPORTED
```
why keep lps support here when it was removed. Commit message states: `- Removed LPS modes`

mnkp replied:
It's probably better not to use a SoC that does not support LPS mode to showcase LPS mode. If anything the example should show that it works also with a SoC that does not support all the power modes.

pizi-nordic replied:
@nordic-krch: No. I think that it is the effect of the development history. I have suggested @ramakrishnapallala, who develops the power management subsystem, to make these definitions generic.

mnkp replied:
> SoC may not support given LPS mode but application may, e.g. dimming screen or LEDs.

I don't quite understand the logic of the sentence. If an application supports SPI transfer but SoC doesn't should the SoC report that it supports SPI transfer?

The main problem with `sys_is_valid_power_state` function is that power subsystem is using it without providing any description (and neither a declaration). Looking at other comments in this PR e.g.
> it looks like this function is generic. It's output depends only on generic kconfig options. Why not move it to the subsystem?

 it looks like we don't really know what this function is supposed to be doing.

The description in zephyr/soc/arm/nordic_nrf/nrf52/soc_power.h says however "Check a low power state is supported by SoC". And in this case the implementation provided in this PR is wrong. We really should clean it up.

pizi-nordic replied:
@mnkp: This is way I asked @ramakrishnapallala, who designed power management, to join this discussion.

nordic-krch commented on soc/arm/nordic_nrf/nrf51/power.c:
```diff
@@ -1,11 +1,96 @@
 
+#define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc);
```
use `LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);` and remove `#define LOG_LEVEL ...`

pizi-nordic replied:
Fixed for nRF51 and nRF52. Thank you.

nordic-krch commented on soc/arm/nordic_nrf/nrf51/power.c:
```diff
@@ -1,11 +1,96 @@
 /*
- * Copyright (c) 2018 Intel Corporation.
+ * Copyright (c) 2017 Intel Corporation.
```
shouldn't you add nordic copyrights after such significant change?

pizi-nordic replied:
This file is almost exact copy of the soc/arm/nordic_nrf/nrf52/power.c. I kept the original copyright of the source.

nordic-krch commented on soc/arm/nordic_nrf/nrf51/power.c:
```diff
@@ -1,11 +1,96 @@
 /*
```
is there any difference between nrf51 and nrf52 implementation? Could that be common?

pizi-nordic replied:
There is subtle (1 line) difference. I have kept the files separate as we do not have "common" folder in `nordic_nrf` three and I would like to not introduce one. I think, that large portions of these files could be moved to global header/source (see PR comment).

pizi-nordic commented:
@dbkinder: Fixed, thank you.

pizi-nordic commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -54,39 +55,38 @@ Running:
-  scheduled and it enters into various Low Power States.
+   *** Power Management Demo on nrf52_pca10040 ***
+   Demo Description
+   Application creates Idleness, Due to which System Idle Thread is
```
Fxied, thank you.

pizi-nordic commented on samples/boards/nrf52/power_mgr/src/main.c:
```diff
@@ -4,56 +4,76 @@
 	}
 
-	printk("\nPress BUTTON1 to enter into Deep Dleep state..."
+	printk("\nPress BUTTON1 to enter into Deep Dleep state. "
```
I can, however I would like to not do so. Reasoning:
1. You cannot automatically wake-up form System OFF, so user must be aware what is happening in order to not perceive system as frozen. So I would like to keep the "Press BUTTON1 ..." procedure.
2. We do not have proper implementation of deep sleep, as wakeup is not handled correctly:
```
config SYS_POWER_DEEP_SLEEP                                                      
        bool "Deep sleep state"                                                  
        depends on SYS_POWER_DEEP_SLEEP_SUPPORTED                                
        help                                                                     
          This option enables the kernel to interface with a power manager       
          application. This permits the system to enter a Deep sleep state       
          supported by the SOC where the system clock is turned off while RAM is 
          retained. This state would be entered when the kernel becomes idle for 
          extended periods and would have a high wake latency.  Resume would be  
          from the reset vector same as cold boot. The interface allows          
          restoration of states that were saved at the time of suspend.          
```
Please note, that we neither retain RAM nor performing proper restoration of peripheral state.
I am going to address these issues in subsequent PRs.

nordic-krch commented on soc/arm/nordic_nrf/nrf51/power.c:
```diff
@@ -1,11 +1,95 @@
+	__enable_irq();
+}
+
+bool sys_is_valid_power_state(enum power_states state)
```
it looks like this function is generic. It's output depends only on generic kconfig options. Why not move it to the subsystem?

pizi-nordic replied:
I am waiting on @ramakrishnapallala opinion.

nordic-krch commented on soc/arm/nordic_nrf/nrf51/soc_power.h:
```diff
@@ -0,0 +1,62 @@
+enum power_states {
+#ifdef CONFIG_SYS_POWER_LOW_POWER_STATE
+# ifdef CONFIG_SYS_POWER_STATE_CPU_LPS_SUPPORTED
+	SYS_POWER_STATE_CPU_LPS,	/* Not used */
```
/* not used */ is kind of misleading. `not supported by SoC`?
Why not keep enum clean without ifdefs?

pizi-nordic replied:
If I do so, user will be not able to add custom power modes to the board and/or application without editing this file.

nordic-krch replied:
it seems that it goes back to moving this enum to the subsystem api file (if `sys_is_valid_power_state` is moved it must go too) then it has to contain all zephyr supported power modes.

pizi-nordic commented on soc/arm/nordic_nrf/nrf51/soc_power.h:
```diff
@@ -0,0 +1,62 @@
+void sys_set_power_state(enum power_states state);
+
+/**
+ * @brief Check a low power state is supported by SoC
```
Thank you

gon1332 commented:
@pizi-nordic Do you build the sample with prj_tickless.conf?
I am porting tte

pizi-nordic commented:
@gon1332:
1. No, I have not tried the tick-less mode.
2. The effect you observe is caused by other thread or timer, which schedules wakeup of the system during main thread sleep. Please review your .config in order to check what features you have enabled.

mnkp commented:
I believe "samples: power: power_mgr: Remove redundant test." commit shouldn't be included in this PR. It removes samples/subsys/power/power_mgr sample application. This is a general test which can run on SoCs from different vendors. It's true that currently only nrf52 series provides power management support but this will change when more SoCs add the feature. I for one, would like to add power management support to SiLabs SoCs.

The two last commits:
- power: Remove sys_is_valid_power_state() interface.
- power: Make power_state enum global.

should be better the first ones introduced by this PR, i.e. should be applied before modification to nrf51, nrf53 power management code is done. But that's being pedantic.

pizi-nordic commented:
It looks that our CI is not working well...

```
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.5.2", minimum required is "3.4") 
-- Selected BOARD qemu_x86_nommu
CMake Error at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/cmake/kconfig.cmake:159 (message):
  command failed with return code: Segmentation fault
Call Stack (most recent call first):
  /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/cmake/app/boilerplate.cmake:284 (include)
  CMakeLists.txt:2 (include)


-- Configuring incomplete, errors occurred!
+ exit 1
```
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/30832/6/console

pizi-nordic commented:
Yes, the work is ongoing. I am trying to push one feature each day :)

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -9,30 +9,31 @@ Overview
 
-This app will cycle through the following power schemes each time idle thread
-calls sys_suspend() hook function :
+This app will cycle through the following power schemes each time system
```
Change to:
```
This app will cycle through the following power schemes each time the system
becomes idle:
```

pizi-nordic replied:
Fixed, thank you.

dbkinder reviewed (changes requested):
Docs are looking better, one more tweak.

mnkp reviewed (dismissed):
Just so my comment is not overlooked:
> I believe we shouldn't remove samples/subsys/power/power_mgr sample application. This is a general test which can run on SoCs from different vendors. It's true that currently only nrf52 series provides power management support but this will change when more SoCs add the feature. 

This PR is getting ever bigger. Maybe we should split it in a few smaller, more manageable parts. E.g. the two following commits could be merged separately:
- power: Remove sys_is_valid_power_state() interface.
- power: Make power_state enum global.

pizi-nordic commented:
>I believe we shouldn't remove samples/subsys/power/power_mgr sample application. This is a general test which can run on SoCs from different vendors. It's true that currently only nrf52 series provides power management support but this will change when more SoCs add the feature.

The samples/subsys/power/power_mgr could be executed on exactly the same hardware and was simply a base for the modified sample. There are other samples, running on other platforms which were mostly kept as is.

I also think that we should have one, generic sample presenting power management, however I do not know how to achieve that - depending on the hardware you will have different power states and different behaviour of each power state.

>This PR is getting ever bigger. Maybe we should split it in a few smaller, more manageable parts. E.g. the two following commits could be merged separately:
>
>     power: Remove sys_is_valid_power_state() interface.
>     power: Make power_state enum global.

I can split this PR, however such action will create PRs which are dependent of each other. From my experience this causes trouble and a lot of effort to keep everything synchronized.

ioannisg reviewed (commented):
@pizi-nordic one quick comment: can you rebase this PR so you get to update also power-mgmt for nRF91 SOC?

dbkinder reviewed (approved):
+1 for doc changes, thanks.

mnkp commented:
> I also think that we should have one, generic sample presenting power management, however I do not know how to achieve that - depending on the hardware you will have different power states and different behaviour of each power state.

Good point, we need to take this into account. But I think that the differences in the behavior of each power state between platforms shouldn't matter. Power Management subsystem defines general meaning of each of the power states. As long as we can enter and exit any given power state that's good enough. Since we have all the `CONFIG_SYS_POWER_STATE_CPU_LPS_SUPPORTED` defines we are able to tailor the testcase to a given platform.

One problem that I see is setting up a wake up event. At the moment power_mgr sample app doesn't do it but merely calls `k_sleep()`. And `k_sleep()` relies on SysTick (for the majority of ARM hardware). In a typical case a device going to low power state would turn off processor clock and along that the SysTick clock. So we would never wake up. Until we add support for low power timers (which work on an independent slow clock) the one common wake up event that I can think of is GPIO interrupt, e.g. push button.

I still think it is better to leave power_mgr sample app in tree and modify it (later) rather than remove completely. Even if it showcases only subset of features provided by the power management subsystem.

> > This PR is getting ever bigger. Maybe we should split it in a few smaller, more manageable parts. E.g. the two following commits could be merged separately:
> > ```
> > power: Remove sys_is_valid_power_state() interface.
> > power: Make power_state enum global.
> > ```
> 
> I can split this PR, however such action will create PRs which are dependent of each other. From my experience this causes trouble and a lot of effort to keep everything synchronized.

This of course would make sense if we could review and merge smaller PRs faster. The two commits I mentioned above are pretty straightforward (along with a few commits that clean up nrf51, nrf52 power management implementation). Now that we start to touch kernel files and modify API the discussion may take longer.

pabigot commented:
FWIW, I think targeting a topic branch with smaller commits would be a good approach for more complex development like this.

pizi-nordic commented:
> Good point, we need to take this into account. But I think that the differences in the behavior of each power state between platforms shouldn't matter. Power Management subsystem defines general meaning of each of the power states. As long as we can enter and exit any given power state that's good enough. Since we have all the CONFIG_SYS_POWER_STATE_CPU_LPS_SUPPORTED defines we are able to tailor the testcase to a given platform.

I have talked with @nashif about that. I will try to make this sample working on non-nordic target.

> One problem that I see is setting up a wake up event. At the moment power_mgr sample app doesn't do it but merely calls k_sleep(). And k_sleep() relies on SysTick (for the majority of ARM hardware). In a typical case a device going to low power state would turn off processor clock and along that the SysTick clock. So we would never wake up. Until we add support for low power timers (which work on an independent slow clock) the one common wake up event that I can think of is GPIO interrupt, e.g. push button.

In case of LPS state, this is problem of the given platform in my opinion. If the policy decided to enter given state and expected wake-up time is different than K_FOREVER, SoC code should arrange the wake-up call.

Do you think, that we should move this responsibility on the APP?

gon1332 commented:
@mnkp GPIO indeed cover the most cases. Another crucial wake up source is the RTC.

The place in which the arrangement of wake up sources has also bothered me in the meantime. The responsibility should be the APP's. As for on how to expose such configuration, my initial thought was to have them done in the configuration code of each peripheral. Eg. In `gpio_pin_configure`, in `rtc_config`.

mnkp commented:
> @mnkp GPIO indeed cover the most cases. Another crucial wake up source is the RTC.

Good point!

> The place in which the arrangement of wake up sources is done, has also bothered me in the meantime. The responsibility should be the APP's. As for on how to expose such configuration, my initial thought was to have them done in the configuration code of each peripheral. Eg. In `gpio_pin_configure`, in `rtc_config`.

We will wake up the device as long as we enable an interrupt. In case of GPIO and RTC APIs it would be enough to set up a callback / alarm. I admit, that's more of an implicit than explicit behavior.

gon1332 commented:
> > The place in which the arrangement of wake up sources is done, has also bothered me in the meantime. The responsibility should be the APP's. As for on how to expose such configuration, my initial thought was to have them done in the configuration code of each peripheral. Eg. In `gpio_pin_configure`, in `rtc_config`.
> 
> We will wake up the device as long as we enable an interrupt. In case of GPIO and RTC APIs it would be enough to set up a callback / alarm. I admit, that's more of an implicit than explicit behavior.

I'm using Zephyr on ST's MCUs and I using ST's Low Power Modes on Zephyr. Although I've made some changes, I'm waiting for this PR to show its final face. As for the RTC, the default driver which configures the RTC in alarm mode, happens to wake up the system even from standby. So, yes, it is an implicit behaviour, that someone has to know.

As for the ST's GPIOs, the initial interrupted configuration are OK to wake you up from a friendly DeepSleep (up to STOP2 mode), but they need to be extended for some special I/O pins that can wake you up from the aggresive DeepSleep modes (Standby, Shutdown). This needs to be declared explicit, somehow.

pizi-nordic committed:
power: Do not suspend devices if device PM is not compiled in

This commit fixes compilation of System Power Management module
when Device Power Management is disabled.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic commented:
I will try to chop this to smaller PR once this will reach the "publishable" state.

pizi-nordic committed:
test: kernel: device.pm: Do not enable system-level PM. It is not used.

This commit disables system-level power management in the device power
management test as it is not used.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

mnkp commented:
> I will try to chop this to smaller PR once this will reach the "publishable" state.

The first part cleans up system power management implementation. Would be nice to have it merged for 1.14.

pizi-nordic committed:
power: Fix naming of Kconfig options controlling low power states

The SYS_POWER_LOW_POWER_STATE_SUPPORTED and SYS_POWER_LOW_POWER_STATE
suggests one low power state but these options control multiple
low power state. This commit uses plural in the names to indicate
that.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
power: Rework OS <-> Application interface

This commit simplifies OS <-> Application interface controlling power
management. In the previous approach application-based PM required
overriding sys_suspend() and sys_resume() functions. As these functions
actually implemented power state change, in such case application
basically had to provide own implementation of all PM-related stuff,
which was not portable and hard to maintain.

This commit changes this scheme: The sys_suspend() and sys_resume()
are now system functions while the application could either use
built-in power management policies or provide its own. All details
of power mode switching are now handled by the OS.

Also, this commit cleans up the Kconfig options related to system-level
power management grouping them under common CONFIG_SYS_PM_ prefix.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
power: Clean up power state names

Some of power states used numerical suffix (_1, _2, etc.) while
otthers not. This commit adds proper suffix to all power state names.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

mnkp commented on kernel/Kconfig.power_mgmt:
```diff
@@ -49,7 +49,7 @@ config SYS_POWER_LOW_POWER_STATE
 	  saving most power.
 
 config SYS_POWER_DEEP_SLEEP
```
Since we are reworking names I think we should use the same convention for both low power and deep sleep. I.e. we should choose one of the pairs:
- `SYS_POWER_LOW_POWER`, `SYS_POWER_DEEP_SLEEP`
- `SYS_POWER_LOW_POWER_STATES`, `SYS_POWER_DEEP_SLEEP_STATES`

mnkp commented on kernel/Kconfig.power_mgmt:
```diff
@@ -49,7 +49,7 @@ config SYS_POWER_LOW_POWER_STATE
 
 config SYS_POWER_DEEP_SLEEP
-	bool "Deep sleep state"
+	bool "Enable Deep Sleep states"
```
"Enable" in Kconfig option is implied, i.e. all the options are there to be enabled. There is no need to be explicit or otherwise all the Kconfig options would start with "Enable ...". It's better to have
```
	bool "Deep Sleep states"
```

mnkp reviewed (commented):
Some of the commits here look like bug fixes. 
- power: Move part of Power Management API to correct group
- power: Fix naming of Kconfig options controlling low power states
- power: Clean up power state names

It would be nice to have them in 1.14.

mnkp commented:
I've noticed only now that there is #13267 addressing some of the issues. It would be great if all power management PRs are labelled with "area: Power Management" otherwise it's easy to miss them.

pizi-nordic committed:
power: Fix naming of Kconfig options controlling deep sleep states

This commit changes the names of SYS_POWER_DEEP_SLEEP* Kconfig
options in order to match SYS_POWER_LOW_POWER_STATE* naming
scheme.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>
